MF-236 - Add support for storing messages in InfluxDB

dborovcanin opened the pull request:
This pull request resolves #236 

Also, this pull request implements service consuming messages from NATS and storing them in InfluxDB.

dborovcanin committed:
Add InfluxDB writer

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Add InfluxDB service to consume normalized messages

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Add all Message fields to influxdb message series

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Add logging and metrics middleware

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Add tests for saving data to Influxdb

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Rename writer to repo

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

anovakovic01 commented on Gopkg.lock:
```diff
@@ -144,6 +144,16 @@
+  ]
+  revision = "02d7d4f043b34ecb4e9b2dbec298c6f9450c2a32"
+  version = "v1.5.2"
+
```
Update `Gopkg.toml` file also.

anovakovic01 commented on cmd/influxdb/main.go:
```diff
@@ -0,0 +1,143 @@
+
+	logger := log.New(os.Stdout)
+	nc, err := nats.Connect(cfg.NatsURL)
+
```
This is nitpicking, but please remove empty line here, in order to group code logically.

anovakovic01 commented on influxdb/influxdb.go:
```diff
@@ -0,0 +1,83 @@
+	"github.com/mainflux/mainflux"
+)
+
+const precision string = "ns"
```
Remove `string`. It is redundant.

anovakovic01 commented on influxdb/influxdb.go:
```diff
@@ -0,0 +1,83 @@
+	client    client.Client
+}
+
+var _ Writer = (*influxRepo)(nil)
```
Place this as first `var`, like in other files.

anovakovic01 commented on influxdb/influxdb_test.go:
```diff
@@ -0,0 +1,112 @@
+		Command:  cmd,
+		Database: testDB,
+	}
+	if response, err := cl.Query(q); err == nil {
```
Extract from `if`, use `err != nil` and remove `else`.

anovakovic01 commented on influxdb/influxdb.go:
```diff
@@ -0,0 +1,83 @@
+}
+
+// New returns new InfluxDB writer.
+func New(cfg client.HTTPConfig, database, pointName string) (Writer, error) {
```
Instead of separating exposed methods and structures with private ones, group them.

anovakovic01 commented on influxdb/influxdb_test.go:
```diff
@@ -0,0 +1,112 @@
+}
+
+// queryDB convenience function to query the database
+func queryDB(cmd string) (res []client.Result, err error) {
```
Place this function at the end, or at the beginning, because it is private helper function.

anovakovic01 commented on influxdb/influxdb_test.go:
```diff
@@ -0,0 +1,112 @@
+
+	res, err := queryDB(q)
+	assert.Nil(t, err)
+	count := res[0].Series[0].Values[0][1]
```
Please place comment here, I don't have any clue what these indices mean.

anovakovic01 commented on influxdb/influxdb.go:
```diff
@@ -0,0 +1,83 @@
+type tags map[string]string
+
+func convertMsg(msg mainflux.Message) (tags, fields) {
+
```
Remove unnecessary empty line.

anovakovic01 commented on cmd/influxdb/main.go:
```diff
@@ -0,0 +1,143 @@
+		logger.Error(fmt.Sprintf("Failed to create InfluxDB writer: %s", err.Error()))
+		return
+	}
+
```
Remove empty line.

dborovcanin committed:
Update Gopkg.toml

Refactor code and update tests.

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

chombium commented on cmd/influxdb/main.go:
```diff
@@ -0,0 +1,142 @@
+	DBPass    string
+}
+
+func handleMsg(logger log.Logger, w influxdb.Writer) nats.MsgHandler {
```
Could you please change the signature of the function handleMsg and use MessageRepository instead of influxdb.Writer. That way this function can be reused in other DB writers.
It would be also nice if the NATS client creation, message subscription and message handling are moved to another separate file or package so that they can be reused by others who write DB writers. This way we will avoid code duplication

mijicd replied:
IMHO, `influxb.Writer` or any kind of writer should be wrapped inside the repository structure. This structure should implement the interface @chombium is talking about. It does not need to be more complex than:

```golang
package mainflux

type MessageRepository interface {
	// Save does some documented magic :)
	Save(Message) error
}
```

drasko commented on influxdb/influxdb.go:
```diff
@@ -0,0 +1,76 @@
+func convertMsg(msg mainflux.Message) (tags, fields) {
+	time := strconv.FormatFloat(msg.Time, 'f', -1, 64)
+	update := strconv.FormatFloat(msg.UpdateTime, 'f', -1, 64)
+	tags := map[string]string{
```
Can you please explain this design - why all these fields are tags? Tags are indexed and time-graphs can be made for one tag, so I think that the `Channel` should be the only tag here.

dborovcanin replied:
Since data tags are used to record metadata, it made sense for me to put all `non-value` data to tags, and value data to `fields`.

drasko replied:
This is not the way InfluxDB works. You need to go through tutorials and understand well the meaning of tags and fields, it is crucial.

This will also have impact on Grafana integration for visualization of messages coming over certain channel.

dborovcanin committed:
Expose HandleMsg at root

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Move test setup to separate file

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Add writers package

Create package `writers` to place all writers implementation there.

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

mijicd commented on writers/README.md:
```diff
@@ -0,0 +1,16 @@
+For an in-depth explanation of the usage of `writers`, as well as thorough
+understanding of Mainflux, please check out the [official documentation][doc].
+
+## Configuration
```
All of these sections are writer-specific, and they should be removed from this file.

mijicd commented on writers/README.md:
```diff
@@ -0,0 +1,16 @@
+# Writers package
+
+Writers package provides an implementation of various `message consumers`.
```
This is not true. Message consumers can consume messages without storing them in any kind of storage. Hence, this should be replaced with `message writers`.

mijicd commented on writers/README.md:
```diff
@@ -0,0 +1,16 @@
+# Writers package
```
Drop the "package" suffix.

mijicd commented on writers/influxdb/influxdb_test.go:
```diff
@@ -0,0 +1,73 @@
+	port      string
+	testLog   = log.New(os.Stdout)
+	testDB    = "test"
+	cl        client.Client
```
Please rename to `client` and change the package alias.

mijicd commented on writers/metrics.go:
```diff
@@ -0,0 +1,30 @@
+
+func (mm *metricsMiddleware) Save(msg mainflux.Message) error {
+	defer func(begin time.Time) {
+		mm.counter.With("method", "handleMessage").Add(1)
```
Use underscore notation to name methods.

mijicd commented on writers/writer.go:
```diff
@@ -0,0 +1,50 @@
+	repo   MessageRepository
+}
+
+func (c *consumer) handleMsg() nats.MsgHandler {
```
Please rename to `consume`. Also, no need for returning the function, just use the same approach as in `normalizer`.

mijicd commented on writers/writer.go:
```diff
@@ -0,0 +1,50 @@
+}
+
+// Start method starts to consume normalized messages received from NATS.
+func Start(name string, nc *nats.Conn, logger log.Logger, repo MessageRepository, counter metrics.Counter, latency metrics.Histogram) error {
```
`name` should be constant, not factory parameter.

dborovcanin replied:
`Name` parameter was meant to be `writer name` or `writer type`, so log messages would make more sense and writer could be distinct from each other.

mijicd commented on cmd/influxdb/main.go:
```diff
@@ -0,0 +1,122 @@
+	DBPass    string
+}
+
+func main() {
```
Consider splitting into multiple helper functions.

mijicd commented on writers/influxdb/influxdb.go:
```diff
@@ -0,0 +1,76 @@
+type fields map[string]interface{}
+type tags map[string]string
+
+func convertMsg(msg mainflux.Message) (tags, fields) {
```
Split into two methods (`tagsOf` and `fieldsOf`) and bind to `influxRepo`.

mijicd commented on writers/influxdb/influxdb.go:
```diff
@@ -0,0 +1,76 @@
+}
+
+// New returns new InfluxDB writer.
+func New(cfg client.HTTPConfig, database, pointName string) (Writer, error) {
```
Instead of passing the configuration, instantiate the client inside `main.go` and do the `defer client.Close()` there. This would remove the entire `Writer` interface and let you rely only on `MessageRepository` as agreed.

mijicd commented on writers/influxdb/messages.go:
```diff
@@ -0,0 +1,11 @@
+import "github.com/mainflux/mainflux/writers"
+
+// Writer represents writer implementation to the InfluxDB.
+type Writer interface {
```
See the remark given above regarding the instantiation of the underlying influxdb client.

dborovcanin committed:
Update README.md

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Refactor NATS message handler

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Use underscore notation in metrics

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Remove Writer interface from influxDB

Factory method change to receive Client instead of client config.
Refactor main method using multiple helper methods.
Update tests and change package aliases.

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Add README.md for InfluxDB writer

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Fix wrong types caused by PR #291

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

mijicd commented on writers/influxdb/README.md:
```diff
@@ -0,0 +1,47 @@
+cd $GOPATH/src/github.com/mainflux/mainflux
+
+# compile the http
+CGO_ENABLED=0 GOOS=linux GOARCH=$GOARCH GOARM=$GOARM go build -ldflags "-s -w" -o ./build/mainflux-influxdb cmd/influxdb/main.go
```
Update `Makefile`.

anovakovic01 replied:
Use `make` here.

mijicd commented on cmd/influxdb/main.go:
```diff
@@ -0,0 +1,138 @@
+	return counter, latency
+}
+
+func main() {
```
Synchronize the structure with `cmd/things/main.go`.

anovakovic01 commented on writers/README.md:
```diff
@@ -0,0 +1,10 @@
+# Writers
+
+Writers provides an implementation of various `message writers`.
```
Fix typo `provide`.

anovakovic01 commented on writers/README.md:
```diff
@@ -0,0 +1,10 @@
+# Writers
+
+Writers provides an implementation of various `message writers`.
+Message consumers are service that consume normalized (in `SenML` format)
```
Fix type `services`.

anovakovic01 commented on writers/influxdb/README.md:
```diff
@@ -0,0 +1,47 @@
+|----------------------------|-----------------------------------|-----------------------|
+| MF_INFLUXDB_DB_WRITER_PORT | Service HTTP port                 | 8180                  |
+| MF_NATS_URL                | NATS instance URL                 | nats://localhost:4222 |
+| MF_INFLUXDB_DB_POINT_NAME  | InfluxDB point to write data to   | messages              |
```
Rename to `MF_INFLUXDB_DB_POINT`.

anovakovic01 commented on writers/influxdb/README.md:
```diff
@@ -0,0 +1,47 @@
+
+| Variable                   | Description                       | Default               |
+|----------------------------|-----------------------------------|-----------------------|
+| MF_INFLUXDB_DB_WRITER_PORT | Service HTTP port                 | 8180                  |
```
Rename to MF_INFLUXDB_WRITER_PORT

anovakovic01 commented on writers/influxdb/README.md:
```diff
@@ -0,0 +1,47 @@
+
+## Usage
+
+Starting service will start consuming messages normalized messages in SenML format.
```
Drop one of two `messages`.

dborovcanin committed:
Refactor main.go to be in sync with other services

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

anovakovic01 commented on writers/influxdb/influxdb.go:
```diff
@@ -0,0 +1,75 @@
+type fields map[string]interface{}
+type tags map[string]string
+
+func (repo *influxRepo) tagsOf(msg *mainflux.Message) tags {
```
Please place private functions and structures at the end.

anovakovic01 commented on writers/influxdb/influxdb_test.go:
```diff
@@ -0,0 +1,78 @@
+		Link:        "link",
+	}
+
+	q := fmt.Sprintf("SELECT * FROM %s", "test..messages\n")
```
Replace with `q := "SELECT * FROM test..messages\n"`.

dborovcanin committed:
Refactor code to be in sync with other services

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Update Makefile to support writer services

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Update README.md to use Make

Remove commented services from Makefile.

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Move consume method at the end of file

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

mijicd commented on Makefile:
```diff
@@ -1,6 +1,7 @@
 BUILD_DIR = build
-SERVICES = users things http normalizer ws
-DOCKERS = $(addprefix docker_,$(SERVICES))
+SERVICES = #users things http normalizer ws
```
? :)

dborovcanin replied:
To speed up testing. :slightly_smiling_face:

mijicd commented on Makefile:
```diff
@@ -28,6 +29,9 @@ proto:
 $(SERVICES): proto
 	$(call compile_service,$(@))
 
+$(WRITERS): proto
```
This task is same as `SERVICES`. Please merge them.

mijicd commented on Makefile:
```diff
@@ -50,6 +54,10 @@ release:
 		docker tag mainflux/$$svc mainflux/$$svc:$(version); \
 		docker push mainflux/$$svc:$(version); \
 	done
+	for svc in $(WRITERS); do \
```
This is the proof for comment above.

dborovcanin committed:
Merge SERVICES and WRITERS tasks in Makefile

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

anovakovic01 reviewed (approved):
LGTM!
